incomplete_cooperative/regret.py

```diff
@@ -0,0 +1,174 @@
+        self.meta_id_to_rank = np.zeros(self.viable_metacoalitions, dtype=int)
+        self.meta_id_to_rank[self.meta_rank_to_id] = np.arange(self.viable_metacoalitions)
+
+        # regret minimizer for every internal node -- except for the bottom layer of game tree
+        self.number_of_regret_minimizers = coalitions_up_to(self.number_of_coalitions, limit_of_revealed - 1)
+
+        self.coalitions_to_player_ids = get_coalition_player_id_map(number_of_players)
+        self.cumulative_regret = np.zeros((self.number_of_regret_minimizers, self.number_of_coalitions), dtype=RMValue)
+        self.cumulative_strategy = np.zeros((self.number_of_regret_minimizers, self.number_of_coalitions),
+                                            dtype=RMValue)
+        self.iteration = 0
+
+    def get_metacoalition_id(self, metacoalition: Iterable[Coalition]) -> int:
+        """Properly get an id of a metacoalition."""
+        coalition_player_ids = self.coalitions_to_player_ids[
+            [x.id for x in metacoalition if len(x) not in [0, 1, 2**self.number_of_players]]]
+        return np.sum(2 ** coalition_player_ids)
+
+    def regret_matching_strategy(self, past_actions: Iterable[Coalition] | int) -> np.ndarray:
+        """Get the action predicted by the strategy using the Regret Matching minimizer.
+
+        If past_actions is an int, it is considered a metacoalition id.
+        Otherwise, it is considered as a list of coalitions of the original game.
+
+        This returns the current prediction, not the averaged strategy over time.
+        """
+        metacoalition = past_actions if isinstance(past_actions, int) else self.get_metacoalition_id(past_actions)
+        rank = self.meta_id_to_rank[metacoalition]
+        # TODO: predictive: přičte predikci ...
+        positive_regret = self.cumulative_regret[rank] * (self.cumulative_regret[rank] > 0)
+        if positive_regret.sum() == 0:
+            positive_regret = np.ones(self.number_of_coalitions)
+            used_coalitions = list(Coalition(metacoalition).players)
+            positive_regret[used_coalitions] = 0
+        return positive_regret / positive_regret.sum()
+
+    def get_average_strategy(self, past_actions: Iterable[Coalition]) -> np.ndarray:
+        """Get average strategy, in terms of actual coalitions in the original game."""
+        metacoalition = self.get_metacoalition_id(past_actions)
+        rank = self.meta_id_to_rank[metacoalition]
+        # in `regret minimization` we only add the "previous" regret matching strategy, not the current one,
+        # so we need to add it here
+        average_strategy = self.cumulative_strategy[rank] + self.regret_matching_strategy(metacoalition)
+        average_strategy = average_strategy / average_strategy.sum()
+        # translate from player_ids to coalition ids
+        return average_strategy[self.coalitions_to_player_ids] * (self.coalitions_to_player_ids > -0.5)
```

Comment:
Why is there the last part? And wouldn't it mess with the normalization?

Reply:
Right, that's a bit of a hack. I'll add a comment about it.

Essentially, in the average strategy we store, we remove the coalitions which are already a part of the K_0.
In the output strategy, we're expecting them to be there again. This line fixes that by indexing with the right ids, where we have positions even for the coalitions from K_0, and they're filled with -1s. So what we get is a re-ordered array, and for every element not in the original array, the element at position -1 is taken. So if we then set the positions where the index is -1 (that's what the multiplication is doing), we set those positions to 0, which is what we want.

Reply:
The values in average_strategy are normalized, and the other values are zero, so it wouldn't mess with the normalization. But just to be sure, I moved the normalization to a later stage.